libfv library module for simple public access to font version string read/write/mod functionality in font-v

chrissimpkins opened the pull request:
Associated issue report https://github.com/source-foundry/font-v/issues/14 from @felipesanches with request for public access to version string read/write/mod functionality in the `font-v` executable.

This PR addresses #14 with a source refactor of the 

- version string read
- version string modification (including version number modification, development/release status indicator modification, and git commit sha1 indicator modification)
- version string write

functionality from the `font-v` executable to a library module `fontv.libfv` with public access to font version string data and methods to achieve the above tasks.

This permits a simple library import to support Python script/executable creation around these tasks outside of this project. The `font-v` executable which will be refactored to use this new library in upcoming releases.

TODO:

- [x] version string reads to `FontVersion` object with automated parsing of data to object properties + unit tests
- [x] version string `DEV` substring modification method + unit tests
- [x] version string `RELEASE` substring modification method + unit tests
- [x] version string `[sha1]` substring modification method + unit tests
- [x] version string `[sha1]-dev` substring modification method + unit tests
- [x] version string `[sha1]-release` substring modification method + unit tests
- [x] version string clear method (removes all metadata from version string) + unit tests
- [x] version string getter method (gets full version string in semicolon delimited format) + unit tests
- [x] version string version number setter method + unit tests
- [x] version string full version string in memory write method (caller can set version string with a string parameter to method) + unit tests
- [x] version string write to font file method (with support for font path rename) + unit tests
- [x] address how to handle version strings that are saved outside of nameID (3,1,1033) and (1,0,0) records - should be exceedingly rare? 
- [x] raise exception if unable to read nameID 5 tables from the font + unit tests
- [x] documentation of libfv library
- [x] add getter methods for FontVersion attributes of use (e.g. status data, SHA1 hash, font path)

chrissimpkins commented:
Still requires the git commit sha1 modification refactor to the libfv library.  Will work on it this week.

Otherwise, the remainder of the planned refactor is complete and unit tests are in place for existing implemented object instantiation and method calls against a combination of possible version string read formats.

Please let me know if you have any thoughts about the approach.

chrissimpkins commented:
git commit sha1 version string mod source added

chrissimpkins commented:
the git commit sha1 modifications are not currently working on Windows.  I think that this may be related to the string that is returned by the `git rev-list` system command on the Windows platform because it is raising IndexError following an attempt to split by newlines based upon the standard output format on OS X and Linux platforms.

chrissimpkins commented:
@texhex are you able to execute the following on a Windows box inside any git repository?  Repo doesn't matter, I just need to see the exact layout of the standard output from this command on Windows to modify the standard output parsing for the git commit sha1 hash (and importantly, that this subcommand is supported on Windows versions of `git.exe`):

```
$ git.exe --abbrev-commit --max-count=1 --format="%h" HEAD
```

Thanks!

chrissimpkins commented:
@jublo reviewed this and he found that Windows git is returning `\n` line endings, we were splitting on `os.linesep` with python and I suspect that this was assuming `\r\n` line endings so the stdout from git python was not being parsed appropriately.  Fixed in commit https://github.com/source-foundry/font-v/pull/15/commits/ce567226890b52be361fc946f8d96874cfd747ef

chrissimpkins commented:
Added otf file tests in https://github.com/source-foundry/font-v/pull/15/commits/4b2a05986962cdf30a616a0bc74776ccee6345ff 

Have complete coverage of the library module with the exception of this block where I still have some questions for discussion:

https://github.com/source-foundry/font-v/pull/15/files#diff-c194b81b78bbb88de24c206806523f7cR82

chrissimpkins commented on lib/fontv/libfv.py:
```diff
@@ -0,0 +1,299 @@
+        # number of sha1 hex characters determined by git (addresses https://github.com/source-foundry/font-v/issues/2)
+        full_git_sha_string = gitpy.rev_list('--abbrev-commit', '--max-count=1', '--format="%h"', 'HEAD')
+        unicode_full_sha_string = tounicode(full_git_sha_string)
+        sha_string_list = unicode_full_sha_string.split("\n")
```
Location of the fix for Windows git sha1 hash parsing issue.  This will need to be fixed in the executable module as well until we transition to use of the library in the font-v executable.

chrissimpkins commented:
@felipesanches library documentation is now available on RTD:

https://font-v.readthedocs.io/en/latest/

felipesanches commented:
Cool, thanks!

I assume I'd better wait for this to be merged before attempting to use the code on fontbakery, right?

chrissimpkins commented:
@davelab6 are you able to comment on the approach to the remaining open bullet point in the OP here (source code linked up there)?  Wondering how important this issue is, whether we should raise an exception for fonts that write to uncommon platformID, platEncID, langID records or let it slide and write back to the record that they used to version their fonts.  I am guessing this is of historic interest only (i.e. used an old font compiler, are not actively compiling new fonts) and unlikely to be encountered during use of a library targeted at font versioning?

Once this is wrapped up I am ready to merge and push this library to PyPI.  Will be available to Felipe when he has time to review for his issue report and we can make changes to address any needs on your end if necessary down the road.

chrissimpkins commented:
@felipesanches if there is anything that you need that is not contained, it would be helpful to know before we merge.  We can always change but I intend to refactor the font-v executable on this library so major breaking changes would be ideal now rather than later if possible.

felipesanches commented:
one quick insight: perhaps we could have a way to access the major and minor version numbers as integers, instead of just the version as a formated string.

felipesanches commented:
Not sure how to deal with "Version 1.001" versus "Version 1.1" though.

Maybe get the version number as a float instead ?

felipesanches commented:
Also, it would be great to parse and store / report ttfautohint metadata. I know that it is not ideal to have that, but currently that's a common practice, so it would be handy to have a helper routine for parsing and preserving it.

chrissimpkins commented:
@felipesanches 

> one quick insight: perhaps we could have a way to access the major and minor version numbers as integers, instead of just the version as a formated string.

The problem is that the OpenType specification allows for leading zeroes and many fonts use this format for versioning.  Is there a way to parse a float in an informative way for the second portion of the version string?

chrissimpkins commented:
> Also, it would be great to parse and store / report ttfautohint metadata. I know that it is not ideal to have that, but currently that's a common practice, so it would be handy to have a helper routine for parsing and preserving it.

Have been giving some thought to this and starting to write up a version string spec over here: https://github.com/source-foundry/font-v/issues/19

My opinion on this is that the ttfautohint metadata does not belong in the font version string.  My feeling is that a developer should always label the version string with the git commit sha1 and this will by definition define their ttfautohint settings assuming that they are hinting with source under version control.  I think that I am going to eliminate this string from Hack and discourage this in the spec here.  Having said that, if Werner sticks with the current substring format we can easily add a method to pull this out of the metadata available in the font.

Thoughts?

felipesanches commented:
in general, by reading your source code, I have the impression that you are not parsing the version data from the TTF upon instantiation of a FontVersion object.

Do you intend to read and parse the data so that users can read the pre-existing fields ? That would be extremely useful. Right now (unless I am mistaken) it seems that the FontVersion class only helps one to set values, and not to read them as they are set on a TTF.

felipesanches commented:
oops, sorry! That's the role of ``_read_version_string`` right ?

chrissimpkins commented:
> Do you intend to read and parse the data so that users can read the pre-existing fields ? That would be extremely useful. Right now (unless I am mistaken) it seems that the FontVersion class only helps one to set values, and not to read them as they are set on a TTF.

No it definitely does set all object properties on instantiation of the FontVersion object from the data in the font binary during the read.  Everything is in memory with all of the modification methods until calling code is ready to write and then you write out any mods with the FontVersion.write_version_string() method (with optional parameter for a new font path if you need it)

chrissimpkins commented:
You can think of the FontVersion object as an in memory representation of the version string in the binary on instantiation.  Everything stays in memory (I don't surprise with any OT table writes, this must be called explicitly) until calling code writes whatever modifications are made to the version string directly through the object properties (e.g. you can modify each of the substring items in the FontVersion.version_string_parts list) or through the public methods.

felipesanches commented:
it could be good to strip() each element after splitting by semicolon on ``_parse_version_substrings``

I'm not sure if anyone pad with white-space but I think that may be common. And if there is white-space we want to get rid of any so we get only the actual data. what do you think ?

chrissimpkins commented:
Currently specifying the possibility of using padding space characters if one wants to use them for improved readability of the version string.  I accounted for this on the parsing of the status string by stripping before an attempt to match on the DEV/RELEASE strings if they are present.  

Maybe a cleanup style formatting method for those who want to remove padding glyphs in the string rather than making this the default?  There may be those who want to maintain the padding.

felipesanches commented:
the most important thing, though, is that I miss an easy way of getting the metadata by type. What I mean is that if 'were delegating parsing inteligence to a FontVersion class, I'd expect the class to provide me a ``get_git_hash`` method that is smart enough to identify which (if any) of the elements of 
self.version_string_parts is a git commit hash and then return me that (or None).

Otherwise, user code could simply split on semicolons without relying on a thirdparty lib

felipesanches commented:
And maybe other similar methods like ``get_version`` (returning just the number, maybe as string, or float or pair of integers, gotta discuss that), ``get_status`` (returning "dev", "release" or None),  ``get_ttfautohint`` (...). And ``get_metadata`` returning a dictionary of it all.

chrissimpkins commented:
> the most important thing, though, is that I miss an easy way of getting the metadata by type.

I really like that.  That was an oversight.  I will include a public method that returns the status substring.

Do you plan to use [git sha1] alone as a substring or as a combination with a `-dev` or `-release` specifier?

felipesanches commented:
Having just the dict would be enough though. No real need for the individual methods if the data is properly tagged in a dict.

chrissimpkins commented:
It somewhat already is a dict but as object properties instead of a formal dictionary.

You can get at version substring with `FontVersion.version` and status substring with `FontVersion.status`

felipesanches commented:
So, here's the deal. Right now we have no idea what's the ideal set of metadata in a font version. ASs discussed previously on https://github.com/googlefonts/fontbakery/issues/1563#issuecomment-346203943 we still have to brainstorm what would be good to have for font projects in general, and for Google Fonts in particular.

felipesanches commented:
> You can get a version substring with FontVersion.version and status substring with FontVersion.status

So we'd need a ``FontVersion.githash`` perhaps... (with None is a hash is not there)

chrissimpkins commented:
The only problem with the githash is that I am not sure how best to match on that.  I can match against `[(0-9)|(a-f)]{7,10}` or something like that but that will potentially pick up non git commit hash strings if there is a short string that uses those characters

chrissimpkins commented:
I was also considering the addition of a method that hashes the entire version string so that you can compare these returned hash strings between two FontVersion objects to determine whether the fonts differ based upon the version strings that they contain.  

Any need for something like that on your end?

felipesanches commented:
As a starting point for the FontBakery test I would simply do something similar to the following:

```python
@register_test
@test(
    id='com.google.fonts/test/166'
  , rationale = """
      The git sha1 tagging and dev/release features of Source Foundry font-v
       tool are awesome and we would love to consider upstreaming the approach
       into fontmake someday. For now we only emit a WARN if a given font does
       not yet follow the experimental versionin style, but at some point we
       may start enforcing it.
    """
  , request = 'https://github.com/googlefonts/fontbakery/issues/1563'
)
def com_google_fonts_test_166(ttFont):
  """ Check for font-v versioning """
  from fontv.libfv import FontVersion

  fv = FontVersion(ttFont.reader.file.name)
  if fv.version and fv.githash and fv.status:
    yield PASS, "Font version string looks GREAT!"
  else:
    yield INFO, ("Version string is:\n{}"
                 "The version string must ideally include a git commit hash"
                 " and eigther a 'dev' or a 'release' suffix such as in the"
                 " example below:\n"
                 " Version 1.3; 0d08353-release").format(fv.version_string)
```

chrissimpkins commented:
> get_status (returning "dev", "release" or None), 

To add to the comment about a public method for this issue, I attempted to implement this here rather than requiring calling code to implement it.

The current approach is that you read in the font to a FontVersion object and then test as follows with object properties:

```python
fv.is_development  # True or False defined
fv.is_release            # True of False defined
```

felipesanches commented:
ok, I updated the example fontbakery code above to address the suggested usage of ``is_dev`` and ``is_release``

chrissimpkins commented:
Would you be willing to modify that code to:

```python
if fv.version and (fv.is_development or fv.is_release):
```

felipesanches commented:
yep, I did :-)

chrissimpkins commented:
is_development and is_release will pick up the git sha1 too, not necessary to include the gitsha1.  Those properties parse on substring 2 containing any of the following:

- `DEV`
- `RELEASE`
- `[sha1]-dev`
- `[sha1]-release`

felipesanches commented:
We might consider forbidding family-names that can be mistaken as git commit hashes LOL.

Or we may require usage of a "git-" preffix on hashes ! (and forbidding families called "git-0d08353-Regualr.ttf" :-P)

felipesanches commented:
In the future there may be other version control systems better and more widespread than git is nowadays, so it may be a safe bet to enforce a version control prefix tag on hashes...

chrissimpkins commented:
Let me add the simple public methods that you recommended and then push this as v0.4.0 so that you can actually dig in a work with it.  As you work out your formatting on your end we can see what holes need to be filled in to support it.

chrissimpkins commented:
> In the future there may be other version control systems better and more widespread than git is nowadays, so it may be a safe bet to enforce a version control prefix tag on hashes...

Not a bad idea.  Let me toss these ideas around a bit tonight and play with the source here.  Will let you know when I have some of the things that you recommended implemented.

chrissimpkins commented:
And could still use feedback from you guys about the nameID 5 records issue that is remains an open bullet point on this PR!

chrissimpkins commented:
Thanks for all of this feedback @felipesanches!  Very helpful!

felipesanches commented:
Regarding this nameId 5 issue you asked about, I'd say you could honor everything, but perhaps report it. Can you add an errors field for reporting this kind of thing perhaps?

If the data is there, I think it should ideally be honored, as long as it is coherent with the rest of the entries (such as being outdated).

Also, it would be great to flag internal inconsistencies, like multiple version metadata with different values in the same font file.

felipesanches commented:
This reminds me of one importante usecase which is to validade the consitency of versioning metadata across several font files in a family. For that, if you could provide an equals operator for the FontVersion class I would be able to transparantly compare several instancies of it for versioning coherence. **That would be ace !** :-D

felipesanches commented:
Gotta run now. I'll be back later (or maybe tomorrow).

Thanks a lot for all the work you've done so far, @chrissimpkins 
I'm getting really happy about this as it seems to be very useful.

chrissimpkins commented:
> This reminds me of one importante usecase which is to validade the consitency of versioning metadata across several font files in a family. For that, if you could provide an equals operator for the FontVersion class I would be able to transparantly compare several instancies of it for versioning coherence. That would be ace ! :-D

I love that idea!

```python
fv1 = FontVersion("font-regular.ttf")
fv2 = FontVersion("font-bold.ttf")
assert fv1 == fv2
```

graphicore commented on lib/fontv/libfv.py:
```diff
@@ -0,0 +1,384 @@
+        self.sha1_develop = sha1_develop
+        self.sha1_release = sha1_release
+
+        self.ttf = ttLib.TTFont(self.fontpath)
```
It would be good if we could `__init__` this by passing an instance of `ttLib.TTFont` instead of `fontpath`. Because I think @felipesanches is planning to use `FontVersion` in Font Bakery, and there we have already `TTFont` instances in memory. (@felipesanches correct me if I'm wrong with my thinking.)

chrissimpkins replied:
How do you think it would be best to include both options (file path and ttLib.TTFont) as part of the instantiation parameters?

```
def __init__(self, fontpath=None, ttfont=None, develop="DEV", release="RELEASE", sha1_develop="-dev", sha1_release="-release"):
```

and then have it explicitly defined by calling code, fail if both remain `None`?

graphicore replied:
I think I remember there was an example somewhere in the wild. I'll try to dig it up now …

chrissimpkins replied:
Thanks Lasse!

chrissimpkins replied:
This should be a simple if/else block around the current assignment of the ttLib.TTFont during instantiation so not a big deal if this is helpful.  This takes place before the remainder of the code to pull the version string in so everything else should stay the same.

graphicore replied:
Having `fontpath` and `ttfont` explicitly is fine for me. 

However, I found what I was reminded of, it's not the same but in the `TTFont` constructor itself, LOL:

https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/ttLib/__init__.py#L65-L72

```
		The constructor can be called with a few different arguments.
		When reading a font from disk, 'file' should be either a pathname
		pointing to a file, or a readable file object
```

Thus `FontVersion` could additionally allow a `file object` as argument.

I would try to go with just one argument and fill the rest with what is there.  But it depends on the semantics of `self.fontpath` (what else is it used for, if not just opening the file, sorry, I didn't read it all to give good advice).

Assuming `fontpath` is mandatory (is it?):
```py
def __init__(self, font, ...):
    try:
        self.fontpath = font.reader.file.name
        self.ttf = font
    except AttributeError: 
        # doesn't look like a TTFont instance with a file
        # expecting a string or a file
        self.ttf = ttLib.TTFont(font)
        # this fails with an AttributeError if font was a file like object but without
        # a name. I'm not sure how important self.fontpath is for the operation
        # of FontVersion
        self.fontpath = self.ttf.reader.file.name
```

chrissimpkins replied:
`fontpath` is only mandatory in the sense that we need to instantiate a `ttLib.TTFont` object and to have the original path to write out changes in the font version string to the original font binary.

I've never used `ttLib.TTFont.reader.file.name`.  This returns the filepath to the font file that was used to instantiate the `TTFont` object?

chrissimpkins replied:
Just confirmed that this is the case, it does maintain the relative path to the font binary.

chrissimpkins replied:
added support for this in https://github.com/source-foundry/font-v/pull/15/commits/e99666012b5e2cb52086aca29680bef32491661b

That works really well Lasse.  Thank you very much!  Will add tests for the new instantiation approach tonight.

graphicore replied:
`reader.file` is like the return value of `file = open(file, "rb")` if opened in `TTFile.__init__` i.e.:

```py
>>> file = open('../fonts/apache/roboto/Roboto-Regular.ttf', 'rb')
>>> file.name
'../fonts/apache/roboto/Roboto-Regular.ttf'
```

But an edge case here is that a file-like object can not have a name attribute, e.g. a [io.BytesIO](https://docs.python.org/2/library/io.html#io.BytesIO):

```py
>>> file1 = open('../fonts/apache/roboto/Roboto-Regular.ttf', 'rb')
>>> file1.name
'../fonts/apache/roboto/Roboto-Regular.ttf'
>>> from io import BytesIO
>>> file2 = BytesIO(file1.read())
# no name!
>>> file2.name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_io.BytesIO' object has no attribute 'name'

>>> from fontTools.ttLib import TTFont

>>> file1.seek(0) # rewind
>>> font1 = TTFont(file)
>>> font1.reader.file.name
'../fonts/apache/roboto/Roboto-Regular.ttf'

# still no name!
>>> font2 = TTFont(file2)
>>> font2.reader.file.name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_io.BytesIO' object has no attribute 'name'

# but it is a good ttFont:
>>> font2['name'].names[0].toUnicode()
u'Copyright 2011 Google Inc. All Rights Reserved.'
```
```

chrissimpkins replied:
Interesting.  Something that would be common enough in any use case to warrant an attempt to handle this?  That seems like an awful lot of work to instantiate the TTFont object when it is as simple as passing a file path string or a file handle.

graphicore replied:
I don't think it needs handling now, it can still be fixed when a real use case comes up.

In the Font Bakery Dashboard, I'd love to have all of Font Bakery running without disk I/O (because the font files are not on the Font Bakery-Worker disks initially, just in memory). But, there are more than enough dependencies that expect reading font files from disk, it won't be any soon that we can run without disk i/o, if ever.

chrissimpkins replied:
Thanks for the suggestion Lasse.  It's now supported and tests are in place for this.  Will part of the initial release of this library.  I really appreciate this feedback.  This was helpful.

chrissimpkins commented:
Added object equality comparisons as recommended by @felipesanches in https://github.com/source-foundry/font-v/pull/15/commits/8d6b05d5c85912f4eba0a6a618d91ec81c99b1e3

![python1](https://user-images.githubusercontent.com/4249591/33403508-abe7fb7e-d52e-11e7-8840-754aaa088590.png)
![python2](https://user-images.githubusercontent.com/4249591/33403511-ae444ae4-d52e-11e7-87ba-7c215fe17981.png)

chrissimpkins commented:
@felipesanches 

> Not sure how to deal with "Version 1.001" versus "Version 1.1" though.

Was giving this some thought last night.  What do you think about splitting the version number on the period into a list, then generating a list of int's from the individual major + minor integers in that string.  This would give you a list of lists like the following for "1.010":

```python
[[1], [0, 1, 0]]
```

If there will be an attempt to automate major/minor version bumps we would need to figure out the code to iterate on a structure like that maybe?

felipesanches commented:
In that case, perhaps ``[1, 0, 1, 0]`` would be preferable. I'm not sure...

The major version number would always be the first integer in the list. All remaining integers would necessarily be ``<= 9``. What do you think ?

felipesanches commented:
Then, a hypotetical ``"Version 14.023"`` would be expressed as ``[14, 0, 2, 3]``.

felipesanches commented:
I think that this could be useful and easy to deal with.

chrissimpkins commented:
OpenType definition allows for more than one major version number.   I've never seen it but it is possible to have `10.10` which wouldn't be unique with that approach.  Could include in a dictionary keyed on 'major', 'minor'?

felipesanches commented:
Maybe a greater than operator for the FontVerion class could be useful as well, so that all of this can be hidden from the user in most usage scenarios.

felipesanches commented:
>  I've never seen it but it is possible to have 10.10 which wouldn't be unique with that approach.

That sounds like it would be expressed as ``[10, 1, 0]`` which is not the same as "Version 1.010" being expressed as ``[1, 0, 1, 0]``

felipesanches commented:
what I mean is that zeroes to the left are only important for the minor version number and because of that the major version number can safely be expresses as a single integer of arbitrarily large value, while minor would require splitting into individual integer "decimal digits".

chrissimpkins commented:
Aaahhhh, I see.  Didn't realize that you meant to keep the full integer value for the major version.

felipesanches commented:
simple enough and effective, right ?

Please consider the ``greater than`` operator, since it will make it trivial to check for version increments (comparing production font files to the current font file under testing)

chrissimpkins commented:
Writing from my phone atm and can't test it.  Do you happen to know if tuple comparisons would work with greater than / less than using (major, minor 1, minor 2, minor 3)?

felipesanches commented:
yes, I guess so. Not sure, though.

felipesanches commented:
please let me know when you publish a new release to pypi so that I can pass on Travis the pytest of the new fontbakery check that uses fontv and then merge this pendind pull request: https://github.com/googlefonts/fontbakery/issues/1563

chrissimpkins commented:
Will test it this evening.  That would be a simple way to compare versions in a greater than / less than fashion if that is something that is useful.

chrissimpkins commented:
> please let me know when you publish a new release to pypi so that I can pass on Travis the pytest of the new fontbakery check that uses fontv and then merge this pendind pull request:

Will do.  I should be able to push this over the weekend.  May add a few of the getter methods that you recommended yesterday and push as is.   Everything else will have open issue reports and we can develop all other outstanding ideas from there.  If you end up using the library on your end, I'm sure that more ideas / problems will emerge based upon your needs.

felipesanches commented:
sure. I'll probably not use all the methods that I mentioned here, though. I only think they are generally useful for users.

chrissimpkins commented:
> sure. I'll probably not use all the methods that I mentioned here, though. I only think they are generally useful for users.

I understand and really appreciate the feedback from both yourself and Lasse.  Use whatever is useful (hopefully some of it is) and thanks for helping to make the rest of it better!

chrissimpkins commented:
Also note the comments from Lasse above that got folded after we addressed them last evening.  You can now instantiate directly with a fontTools.ttLib.TTFont object in addition to a filepath.  He wanted support for this in order to avoid additional file reads when you are already working with a fontTools object.

Also in case you missed it, your equality and inequality comparisons are included now! :)

felipesanches commented:
Ah! Great!

Yeah, I had wondered about that same thing (passing a TTFont instead of path) but forgot to mention

chrissimpkins commented:
Added in https://github.com/source-foundry/font-v/pull/15/commits/18cd869359291fa30951bd5bc3c7c78347347377 based upon your comments above:

- `FontVersion.get_version_number_tuple()` - returns tuple of (major version, minor version 1, minor version 2, minor version 3) as integer items
- `FontVersion.get_metadata_list()` - returns list of metadata string items
- `FontVersion.get_status_substring()` - returns the status substring

Also added commented out source for greater than / less than comparisons with the FontVersion object.  I am going to push this back to a future release because I want to make sure I fully understand the result of the tuple comparisons by values such as `(1, 0, 0, 1)` vs. `(1, 0, 1, 0)` and `(1, 1)` vs. `(1, 0, 0, 1)`.  These will differ but is that what we want?  My thought is yes but will give this more thought before implementation.

Will push this as release v0.4.0 this weekend after I update the readme documentation.  @felipesanches will let you know when this is available on PyPI so that you can update your install approach on Font Bakery.

Thanks again to both of you for all of your help!

chrissimpkins commented:
merging to dev branch to do some work on the font-v executable.

chrissimpkins commented:
@felipesanches this will be a transient branch and will be deleted once I can confirm that you are no longer installing directly from it.  Let's plan to communicate about this issue once I push the release so that I don't eliminate it while you still depend on it for the install.

felipesanches commented:
feel free to delete the branch. Travis will fail for me and I'll move on to installing directly via PyPI ;-)

chrissimpkins commented:
@felipesanches was hoping to avoid causing your tests to fail as the notification :)

felipesanches commented:
heh! No worries about that :-D
